src/tools/pkcs11-tool.c

```diff
@@ -8293,6 +8303,146 @@ static const char *p11_utf8_to_local(CK_UTF8CHAR *string, size_t len)
 	return buffer;
 }
 
+static CK_BBOOL
+p11_is_percent_format_reserved_char(CK_UTF8CHAR c)
+{
+	switch (c) {
+	case ' ':
+	case '!':
+	case '"':
+	case '#':
+	case '$':
+	case '%':
+	case '&':
+	case '\'':
+	case '(':
+	case ')':
+	case '*':
+	case '+':
+	case ',':
+	case '/':
+	case ':':
+	case ';':
+	case '=':
+	case '?':
+	case '@':
+	case '[':
+	case ']':
+		return CK_TRUE;
+	}
+	return CK_FALSE;
+}
+
+static char
+p11_percent_format_first(CK_UTF8CHAR c)
+{
+	switch (c) {
+	case ' ':
+	case '!':
+	case '"':
+	case '#':
+	case '$':
+	case '%':
+	case '&':
+	case '\'':
+	case '(':
+	case ')':
+	case '*':
+	case '+':
+	case ',':
+	case '/':
+		return '2';
+	case ':':
+	case ';':
+	case '=':
+	case '?':
+		return '3';
+	case '@':
+		return '4';
+	case '[':
+	case ']':
+		return '5';
+	}
+	return c;
+}
+
+static char
+p11_percent_format_second(CK_UTF8CHAR c)
+{
+	switch (c) {
+	case ' ':
+		return '0';
+	case '!':
+		return '1';
+	case '"':
+		return '2';
+	case '#':
+		return '3';
+	case '$':
+		return '4';
+	case '%':
+		return '5';
+	case '&':
+		return '6';
+	case '\'':
+		return '7';
+	case '(':
+		return '8';
+	case ')':
+		return '9';
+	case '*':
+		return 'A';
+	case '+':
+		return 'B';
+	case ',':
+		return 'C';
+	case '/':
+		return 'F';
+	case ':':
+		return 'A';
+	case ';':
+		return 'B';
+	case '=':
+		return 'D';
+	case '?':
+		return 'F';
+	case '@':
+		return '0';
+	case '[':
+		return 'B';
+	case ']':
+		return 'D';
+	}
+	return c;
+}
+
+static const char *
+p11_utf8_to_local_percent_format(CK_UTF8CHAR *string, size_t len)
+{
+	static char buffer[1024];
+	size_t output_index, input_index;
+
+	while (len && string[len - 1] == ' ')
+		len--;
+
+	for (output_index = input_index = 0; output_index < sizeof(buffer) - 1;
+			output_index++) {
+		if (input_index >= len) {
+			break;
+		}
+		if (p11_is_percent_format_reserved_char(string[input_index])) {
+			buffer[output_index++] = '%';
+			buffer[output_index++] = p11_percent_format_first(string[input_index]);
+			buffer[output_index] = p11_percent_format_second(string[input_index]);
```

Comment:
> This can also go over the buffer limits as the `output_index` is only checked to be in bounds for the first line, but not after the `output_index` is incremented twice without checking here.

You are completely right. I will update the maximum value. Nice catch

Reply:
> I think this is just the hexadecimal value of the ascii character. Isn't there a simple way to do this, such as converting the character to number and doing some bit masks or just doing the `printf` with `%X` format modifier?

I will investigate it. Thanks for the tip